### PR TITLE
IAP for system services

### PR DESCRIPTION
**Pull request by kron4eg**
**What this PR does / why we need it**:
Protect system resources during outside access with OIDC

**Special notes for your reviewer**:

**Documentation**:
https://github.com/kubermatic/docs/pull/61

**Release note**:
```release-note
Breaking change: basic auth for kibana/grafana/prometheus/alertmanager replaced with oAuth
```

### CODE REVIEW

**Comment by mrIncompetent** [comment]
Please set the PR description according to the template

File: config/iap/templates/deployments.yaml [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+apiVersion: apps/v1
+kind: Deployment
+metadata:
+  name: {{ template "iap.fullname" $ }}-{{ .name }}
+  labels:
+    app: {{ template "iap.name" $ }}
+    chart: {{ template "iap.chart" $ }}
+    release: {{ $.Release.Name }}
+    heritage: {{ $.Release.Service }}
+    target: {{ .name }}
```
Comment (mrIncompetent) [inline config/iap/templates/deployments.yaml]: Do we need all of this?
Our charts so far have been fairly minimal with focus on simplicity

File: config/iap/templates/deployments.yaml [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+spec:
+  replicas: 1
```
Comment (mrIncompetent) [inline config/iap/templates/deployments.yaml]: Should be configurable

File: config/iap/templates/_helpers.tpl [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+{{- end -}}
+{{- end -}}
+{{- end -}}
+
+{{/*
+Create chart name and version as used by the chart label.
+*/}}
+{{- define "iap.chart" -}}
+{{- printf "%s-%s" .Chart.Name .Chart.Version | replace "+" "_" | trunc 63 | trimSuffix "-" -}}
+{{- end -}}
```
Comment (mrIncompetent) [inline config/iap/templates/_helpers.tpl]: Do we need all of this?
Our charts so far have been fairly minimal with focus on simplicity

File: config/iap/templates/ingresses.yaml [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+apiVersion: extensions/v1beta1
+kind: Ingress
+metadata:
+  name: {{ .name }}-iap
+  labels:
+    app: {{ template "iap.name" $ }}
+    chart: {{ template "iap.chart" $ }}
+    release: {{ $.Release.Name }}
+    heritage: {{ $.Release.Service }}
+    target: {{ .name }}
```
Comment (mrIncompetent) [inline config/iap/templates/ingresses.yaml]: Do we need all of this?
Our charts so far have been fairly minimal with focus on simplicity

File: config/iap/templates/services.yaml [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+apiVersion: v1
+kind: Service
+metadata:
+  name: {{ .name }}-iap
+  labels:
+    app: {{ template "iap.name" $ }}
+    chart: {{ template "iap.chart" $ }}
+    release: {{ $.Release.Name }}
+    heritage: {{ $.Release.Service }}
+    target: {{ .name }}
```
Comment (mrIncompetent) [inline config/iap/templates/services.yaml]: Do we need all of this?
Our charts so far have been fairly minimal with focus on simplicity

File: config/monitoring/alertmanager/values.yaml [hunk 6]
```diff
@@ -1,7 +1,5 @@
 alertmanager:
-  version: v0.15.0
-  # admin:loodse123
-  auth: 'YWRtaW46JGFwcjEkUFR0Y3lQc0MkWVZFTk9nZmpiVE52RWtSeVk1ZjJULgo='
+  version: v0.15.2
```
Comment (mrIncompetent) [inline config/monitoring/alertmanager/values.yaml]: Should be done in a dedicated PR

**Review by mrIncompetent — CHANGES_REQUESTED** [review]

File: config/iap/templates/deployments.yaml [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+---
+apiVersion: apps/v1
+kind: Deployment
+metadata:
+  name: {{ template "iap.fullname" $ }}-{{ .name }}
+  labels:
+    app: {{ template "iap.name" $ }}
+    chart: {{ template "iap.chart" $ }}
+    release: {{ $.Release.Name }}
+    heritage: {{ $.Release.Service }}
```
Comment (pkavajin) [inline config/iap/templates/deployments.yaml]: Why do these labels have to specified manually? Shouldnt helm do it on itself?

**Comment by mrIncompetent** [comment]
Documentation PR is missing

File: config/iap/templates/deployments.yaml (hunk 7 above)
Comment (kron4eg) [reply config/iap/templates/deployments.yaml]: nope, it's helm convention, have to provide them manually

File: config/iap/templates/deployments.yaml (hunk 2 above)
Comment (kron4eg) [reply config/iap/templates/deployments.yaml]: I'm not sure if it's possible (and have any sense). By default keycloack-proxy maintain it's state (tokens) in-memory boltdb. To have `replicas >1`, we'd need a redis to save tokens there, which I've avoided. Also system services like those wouldn't experience huge loads on them.

**Comment by kron4eg** [comment]
@mrIncompetent all fixed

**Comment by mrIncompetent** [comment]
@kron4eg 
PR looks good!
Could you write down the steps needed for us to limit access to our Github org?
So we don't expose all those on cloud.kubermatic.io to everyone

**Review by pkavajin — DISMISSED** [review]

**Comment by kron4eg** [comment]
@mrIncompetent that's very good question :)

By looking at the docs I see that we would need to use combination of dex and keycloak-proxy configs.

on keycloak-proxy side:
* `--scopes=groups+federated:id` (ask dex to supply custom claim to the answer token)
* `--match-claims=federated_claims:connector_id=internal_github`
* `--match-claims=groups=kubermatic:developers` (optional)
* `--resources "uri=/*|groups=kubermatic:developers"` (optional and alternative to previous match-claim)

on dex side:
* add connector type github, but with ID `internal_github`, which will require
```
      orgs:
      - name: kubermatic
```

**Comment by kron4eg** [comment]
such stunt however will require supporting (via values.yaml) additional flags for keycloack-proxy, and thus require changing this PR

**Comment by kron4eg** [comment]
@mrIncompetent so here is combination that I've got working

```
scopes:
- "groups"
resources:
- uri: "/*"
  groups:
  - "mygithuborg:dev"
```

And it restricts access to that one organization, even though dex itself can permit everyone from github. But we will need second `internal_github` anyway, since dex is not supplying groups clams if orgs are not specified in dex's connector config

**Comment by kron4eg** [comment]
@mrIncompetent currently keycloak-proxy is unable to match claims that are hashmaps (like dex issuing). I've opened upstream issue for this: 
https://github.com/gambol99/keycloak-proxy/issues/413

File: config/monitoring/grafana/values.yaml [hunk 8]
```diff
@@ -1,7 +1,6 @@
 grafana:
   user: YWRtaW4= # admin
```
Comment (pkavajin) [inline config/monitoring/grafana/values.yaml:2]: isnt that obsolete now by oauth identities?

File: config/monitoring/grafana/config/grafana.ini [hunk 9]
```diff
@@ -1,3 +1,15 @@
 [security]
```
Comment (xrstf) [inline config/monitoring/grafana/config/grafana.ini:1]: Why do we still need to configure username and password if the login form and basic auth are disabled?

Comment (xrstf) [reply config/monitoring/grafana/config/grafana.ini:1]: > for backward compatibility
> I’m going to remove it after full IAP deployment

**Review by xrstf — DISMISSED** [review]
LGTM

**Review by pkavajin — DISMISSED** [review]

**Review by mrIncompetent — DISMISSED** [review]

**Review by mrIncompetent — APPROVED** [review]

**Pull request merged**